// proto3記法であることを示す
syntax = "proto3";

package todo;


/*
message命名規則(ベストプラクティスはどんなものだろう……)
*Component : Request/Responseに用いるコンポーネント
Todo[Action]Request : Action用リスエスト内容
Todo[Action]Response : Action用レスポンス内容
*/

// TODO
message TodoComponent {
    string todo_name = 1;
    bool   is_done = 2;
}

// サーバーからのレスポンス(共通)
message ServerResponseComponent {
    bool    is_success = 1;
    string  message = 2;
}

// TODOを作成するときのリクエスト
message TodoCreateRequest {
    string todo_name = 1;
    string timestamp = 3;
}

// TODOを作成するときのレスポンス
// messageの中に別のmessageを含むことができる
message TodoCreateResponse {
    ServerResponseComponent response = 1;
    string                  timestamp = 2;
}

message TodoShowRequest {
    string timestamp = 1;
}

// repeated SOMETHING = SOMETHING[]
message TodoShowResponse {
    repeated TodoComponent  todos = 1;
    string                  timestamp = 3;
}

message TodoUpdateRequest {
    TodoComponent   todo = 1;
    string          timestamp = 2;
}

message TodoUpdateResponse {
    ServerResponseComponent response = 1;
    string                  timestamp = 2;
}

// ルーティング
service TodoGateway {
    rpc TodoCreate (TodoCreateRequest) returns (TodoCreateResponse) {}
    rpc TodoShow (TodoShowRequest) returns (TodoShowResponse) {};
    rpc TodoUpdate (TodoUpdateRequest) returns (TodoUpdateResponse) {}
}
